src/sinks/util/compressor.rs

```diff
@@ -17,21 +22,67 @@ impl Writer {
     pub fn get_ref(&self) -> &BytesMut {
         match self {
             Writer::Plain(inner) => inner.get_ref(),
-            Writer::Gzip(inner) => inner.get_ref().get_ref(),
-            Writer::Zlib(inner) => inner.get_ref().get_ref(),
+            Writer::Gzip(inner) => inner.get_ref().get_ref().get_ref(),
+            Writer::Zlib(inner) => inner.get_ref().get_ref().get_ref(),
             Writer::Zstd(inner) => inner.get_ref().get_ref(),
             Writer::Snappy(inner) => inner.get_ref().get_ref(),
         }
     }
+
+    pub fn into_inner(self) -> BytesMut {
+        match self {
+            Writer::Plain(writer) => writer,
+            Writer::Gzip(writer) => writer
+                .into_inner()
+                .expect("BufWriter writer should not fail to finish")
+                .finish()
+                .expect("gzip writer should not fail to finish"),
+            Writer::Zlib(writer) => writer
+                .into_inner()
+                .expect("BufWriter writer should not fail to finish")
+                .finish()
+                .expect("zlib writer should not fail to finish"),
+            Writer::Zstd(writer) => writer
+                .finish()
+                .expect("zstd writer should not fail to finish"),
+            Writer::Snappy(writer) => writer
+                .finish()
+                .expect("snappy writer should not fail to finish"),
+        }
+        .into_inner()
+    }
+
+    pub fn finish(self) -> io::Result<BytesMut> {
+        let buf = match self {
+            Writer::Plain(writer) => writer,
+            Writer::Gzip(writer) => writer.into_inner()?.finish()?,
+            Writer::Zlib(writer) => writer.into_inner()?.finish()?,
+            Writer::Zstd(writer) => writer.finish()?,
+            Writer::Snappy(writer) => writer.finish()?,
+        }
+        .into_inner();
+
+        Ok(buf)
+    }
 }
 
 impl From<Compression> for Writer {
     fn from(compression: Compression) -> Self {
-        let writer = BytesMut::with_capacity(1_024).writer();
+        let writer = BytesMut::with_capacity(OUTPUT_BUFFER_CAPACITY).writer();
         match compression {
             Compression::None => Writer::Plain(writer),
-            Compression::Gzip(level) => Writer::Gzip(GzEncoder::new(writer, level.as_flate2())),
-            Compression::Zlib(level) => Writer::Zlib(ZlibEncoder::new(writer, level.as_flate2())),
+            // Buffering writes to the underlying Encoder writer
+            // to avoid Vec-trashing and expensive memset syscalls.
```

Comment:
Thanks for the note here. Could you also add a link to https://github.com/rust-lang/flate2-rs/issues/395#issuecomment-1975088152 just to pull in that context?

Reply:
Done.